### PR TITLE
CSV normalize names option

### PR DESCRIPTION
**Pull request by tdoehmen**
This PR relates to the discussion in PR #1789. A boolean option "normalize_names" (default: FALSE) was added to the read_csv(_auto) function, which will normalize all column names in such a way that they can be used in an SQL statement without quotes. Following transformations are applied to the raw input:
1. all characters are converted to lowercase
2. all accents are removed
3. all non 0-9 and a-z characters are removed
4. if the name is an SQL keyword or starts with a number, a _ will be prepended

The PR also contains a simplified, more robust, and more efficient procedure to uniquify column names, resulting in slight changes to the way name collisions are handled. If there are multiple columns with the same name, the first will keep its name, and the proceeding ones will receive an increasing number, starting at 1 (without zero-padding). Example:
a,a,a,a,a,a,a,a,a,a,a,a -> a,a_1,a_2,...a_11 (previously: a_00,a_01,a_02,...,a_11)

### CODE REVIEW

File: src/execution/operator/persistent/buffered_csv_reader.cpp
Comment (Mytherin) [inline src/execution/operator/persistent/buffered_csv_reader.cpp:181]: Perhaps remove accents before doing the lower-case conversion? The `StringUtil::Lower` is not unicode aware, so otherwise you might have `Á` -> `A` and end up with an uppercase character in the final name anyway.

The reason this seems to work in the example below is that DuckDB can do lowercase -> uppercase matching now as long as there are no ambiguities, e.g.:

```sql
-- table with uppercase name
create table integers("HELLO" integer);
insert into integers values (1);
-- this works!
select hello from integers;
-- print the actual name
select alias(hello) from integers;
-- HELLO
```

With ambiguities this is not allowed though:
```sql
-- hello with different cases
create table integers2("HELLO" integer, "hEllO" integer);
-- error! ambiguity between HELLO and hEllO
select hello from integers2;
-- OK! match case exactly
select "HELLO" from integers2;
```

File: test/sql/copy/csv/data/auto/normalize_names_4.csv
Comment (Mytherin) [inline test/sql/copy/csv/data/auto/normalize_names_4.csv:1]: Could you add a test with spaces in the names, e.g. `  a  , b  ,  c  `?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great, and very nice feature. Some comments:

File: src/execution/operator/persistent/buffered_csv_reader.cpp
Comment (tdoehmen) [reply src/execution/operator/persistent/buffered_csv_reader.cpp]: Thanks, I would say then it should even be done after throwing out all UTF-8 characters. Did this now!

File: test/sql/copy/csv/data/auto/normalize_names_4.csv
Comment (tdoehmen) [reply test/sql/copy/csv/data/auto/normalize_names_4.csv:1]: Done!

**Comment by Mytherin** [comment]
Great, thanks! The code looks good now and is ready to merge after the tests pass.

One more point of discussion: do we want to (partly) enable this by default? For example, at least stripping spaces seems like a good idea to do by default since column names with spaces in them are not very fun (and likely unintentional).

**Comment by tdoehmen** [comment]
Do we talk about removing all whitespaces or about leading/trailing whitespaces? With the latter, I would agree. It's indeed frustrating because you can't really "see" them. For whitespaces in between words (idk. "Experiment a", "Experiment b") I am not so sure. I would tend to say it's better to leave them by default to maintain the original meaning. 

This made me realize something which is maybe not so pretty about the current PR. The "normalize_names" option removes all spaces, also in between words. Maybe it would be more sensible to replace those by _ actually. If I make another push, I would change that in the same go.

**Comment by hawkfish** [comment]
I was actually surprised to see that the spec calls for leading and trailing spaces to be included in column names. The Tableau reader stripped them, and no one ever complained. I think that the old Microsoft Jet driver did this too, so I'd say go with stripping.

**Comment by Mytherin** [comment]
Perhaps we could have three different normalization modes, "auto" (the default) that strips leading and trailing whitespace and perhaps replaces spaces with underscores, "full/true/on" which does full normalization as implemented in this PR, and "none/false/off" which performs no normalization and leaves the column names as they appear in the file.

**Comment by danz68** [comment]
I might be late, yet here are my thoughts about this matter.
²
Any CSV header requires at least a validity check, if not a transformation. IMO, the default transformations, if required, are: 
1. trailing space (and not whitespace) should be removed by default (related to how SQL treats CHAR values equalities, ‘A     ‘=  ‘A’ is true, but only for regular space char, ascii 32)
2*. blank field names require consistent naming/unique identification per table. 
3*. duplicate field names require renaming/unique identification per table. 

Optional transformations
4. any leading space or whitespace to be removed
5. any trailing whitespace to be removed
6. any accent to be removed (not sure how is implemented in duckdb but I expect to be equivalent with the transformation of the character into the closest ASCII shape/visual identification,  “m²” to be replaced by “m2”)
7. any space/whitespace to be replaced by underscore (or dot or anything else)
8. other common characters like brackets to be replaced by underscore or anything else, or removed. To be honest "Qty (m²)" looks just right, "qty__m2_" not so, "qty_m2" is acceptable.  

* Always points 2 and 3 need to be the last two checks/transformations.

**Comment by Mytherin** [comment]
@danz68 agreed these points all seem reasonable. I think that is very similar to what we are doing in this PR actually.

**Comment by tdoehmen** [comment]
Okay, I added the trimming per default now and switched to using NFKD normalization to e.g. support the "Qty (m²)" -> "qty_m2" conversion. The parameter I left as boolean for now - it shouldn't hurt anyone when leading/trailing whitespaces are being removed by default. If this turns out to be problematic for anyone or if the normalize_names=TRUE turns out to be used frequently by many, we can still change the defaults and/or offer more subtle degrees of normalization here.

**Review by Mytherin — APPROVED** [review]
Thanks for the changes! This looks great now. I think this is ready to merge after the tests pass.

**Comment by Mytherin** [comment]
Could you run the format-fix please? After that it is ready to merge.

**Comment by tdoehmen** [comment]
I've had some problems with the format scripts on windows recently. One minor problem is that it uses `os.rename` to copy the `'.tmp'` files, which doesn't work on windows if the file already exists. That's easy to fix by replacing it with `os.replace`. But the main problem is that the formatting script introduces an additional CR LF after each line. I mean it looks pretty, but it doesn't seem like it's intentional as I haven't seen any source files in the project which have this.

**Comment by Mytherin** [comment]
Hm, no that isn't intentional. This might be down to a split happening on `\n` which could leave dangling carriage returns if the files are partitioned by windows newlines (`\r\n`). Perhaps you could search the `format.py` script and replace the occurrences of `.split('\n')` with `.replace('\r\n', '\n').split('\n')`?

**Comment by tdoehmen** [comment]
It fixed the check but not the automatic format fixing. But I think this suffices as solution for me. Format check seems to pass now.

**Comment by hawkfish** [comment]
@tdoehmen Do you have your Windows editor set to use LF? I've found that some Win editors can be quite unhelpful about this.  We also have an `.editorconfig` file that can make this easier if your editor can read it.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**
